flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/StringToBinaryCastRule.java

```diff
@@ -45,6 +52,28 @@ public String generateExpression(
             String inputTerm,
             LogicalType inputLogicalType,
             LogicalType targetLogicalType) {
-        return methodCall(inputTerm, "toBytes");
+        int inputLength = Integer.MAX_VALUE;
+        if (inputLogicalType instanceof CharType) {
+            inputLength = ((CharType) inputLogicalType).getLength();
+        } else if (inputLogicalType instanceof VarCharType) {
+            inputLength = ((VarCharType) inputLogicalType).getLength();
+        }
+
+        int targetLength = Integer.MAX_VALUE;
+        if (targetLogicalType instanceof BinaryType) {
+            targetLength = ((BinaryType) targetLogicalType).getLength();
+        } else if (targetLogicalType instanceof VarBinaryType) {
+            targetLength = ((VarBinaryType) targetLogicalType).getLength();
+        }
+        targetLength = Math.min(targetLength, inputLength);
```

Comment:
I would remove that, and instead below have a check like:
```
if (targetLength >= inputLength) {
    return methodCall(inputTerm, "toBytes");
} else {
    return staticCall(
            Arrays.class, "copyOfRange", methodCall(inputTerm, "toBytes"), 0, targetLength);
}
```

Reply:
Hey @matriv , thanks for your review! I will update the code here.